mlflow/spark.py

```diff
@@ -103,9 +107,36 @@ def log_model(spark_model, artifact_path, conda_env=None, jars=None, dfs_tmpdir=
     >>> model = pipeline.fit(training)
     >>> mlflow.spark.log_model(model, "spark-model")
     """
-    return Model.log(artifact_path=artifact_path, flavor=mlflow.spark, spark_model=spark_model,
-                     jars=jars, conda_env=conda_env, dfs_tmpdir=dfs_tmpdir,
-                     sample_input=sample_input)
+    _validate_model(spark_model, jars)
+    run_id = mlflow.tracking.fluent._get_or_start_run().info.run_uuid
+    run_root_artifact_uri = mlflow.get_artifact_uri()
+    # If the artifact URI is a local filesystem path, defer to Model.log() to persist the model
+    if mlflow.tracking.utils._is_local_uri(run_root_artifact_uri):
+        return Model.log(artifact_path=artifact_path, flavor=mlflow.spark, spark_model=spark_model,
+                         jars=jars, conda_env=conda_env, dfs_tmpdir=dfs_tmpdir,
+                         sample_input=sample_input)
+    # If Spark cannot write directly to the artifact repo, defer to Model.log() to persist the
+    # model
+    model_dir = os.path.join(run_root_artifact_uri, artifact_path)
+    try:
+        spark_model.save(os.path.join(model_dir, _SPARK_MODEL_PATH_SUB))
+    except Py4JJavaError:
+        return Model.log(artifact_path=artifact_path, flavor=mlflow.spark, spark_model=spark_model,
+                         jars=jars, conda_env=conda_env, dfs_tmpdir=dfs_tmpdir,
+                         sample_input=sample_input)
+
+    # Otherwise, override the default model log behavior and save model directly to artifact repo
+    mlflow_model = Model(artifact_path=artifact_path, run_id=run_id)
+    with TempDir() as tmp:
+        tmp_model_metadata_dir = tmp.path("model")
```

Comment:
Correct me if I'm wrong, but it seems like we could just do `with TempDir() as tmp_model_metadata_dir` and avoid creating a new `tmp.path("model")` subdirectory.

Reply:
Funnily enough, `TempDir()` returns a Python object & the only way to get a path out of it is to call `path()`. I'd just copied this code from the existing `Model.log` implementation ([link](https://github.com/mlflow/mlflow/blob/master/mlflow/models/__init__.py#L70)) - I might keep it the same for now & just add a comment referencing the copied code.

Reply:
Can we do `tmp.path()` instead of `tmp.path("model")` + `os.makedir()`. It's simpler.